src/modules/auto-state-on-scroll-bounds.js

```diff
@@ -0,0 +1,178 @@
+	 	<add data-x-start-state-rem-class="" />
+	 	<add data-x-end-state-add-class="" />
+	 	<add data-x-end-state-rem-class="" />
+	 	<add data-y-start-state-add-class="" />
+	 	<add data-y-start-state-rem-class="" />
+	 	<add data-y-end-state-add-class="" />
+	 	<add data-y-end-state-rem-class="" />
+ 	
+ 	Content: Element that is inside the container and that contains all of the content.
+ 			 Will be used to evaluate total scrolling width.
+ 		<add class="js-scroll-bounds" />
+ */
+(function ($, undefined) {
+	
+	'use strict';
+	var win = $(window);
+	var site = $('#site');
+	var scrollTimer = 0;
+	var resizeTimer = 0;
+	
+	var elements = $();
+	var defaultThreshold = 0;
+	
+	var SELECTOR_CTN = '.js-scroll-bounds-ctn';
+	var SELECTOR_SCROLL = '.js-scroll-bounds';
+	
+	var ATTR_THRESHOLD = 'data-scroll-bounds-threshold';
+	
+	var getBounds = function (elems) {
+		var b = {
+			x: 0,
+			y: 0
+		};
+		
+		elems.each(function () {
+			var t = $(this);
+			var x = t.outerWidth() + t.offset().left;
+			var y = t.outerHeight() + t.offset().top;
+			
+			b.x = Math.max(x, b.x);
+			b.y = Math.max(y, b.y);
+		});
+		
+		b.x = Math.floor(b.x);
+		b.y = Math.floor(b.y);
+		
+		return b;
+	};
+	
+	var computeStates = function (elem) {
+		var ctn = elem.closest(SELECTOR_CTN);
+		var curX = Math.floor(elem.scrollLeft());
+		var curY = Math.floor(elem.scrollTop());
+		var elemW = Math.floor(elem.outerWidth());
+		var elemH = Math.floor(elem.outerHeight());
+		var scrollW = Math.floor(elem.prop('scrollWidth'));
+		var scrollH = Math.floor(elem.prop('scrollHeight'));
+		var isScrollableX = scrollW > elemW;
+		var isScrollableY = scrollH > elemH;
+		var thresholdValue = !!elem.attr(ATTR_THRESHOLD) ?
+			parseFloat(elem.attr(ATTR_THRESHOLD)) : defaultThreshold;
+		var xThreshold = Math.floor(scrollW * thresholdValue);
+		var yThreshold = Math.floor(scrollH * thresholdValue);
+		
+		
+		var xStartVisible = !isScrollableX || curX <= xThreshold;
+		var xEndVisible = !isScrollableX || (scrollW - (curX + elemW)) <= xThreshold;
+		var yStartVisible = !isScrollableY || curY <= yThreshold;
+		var yEndVisible = !isScrollableY || (scrollH - (curY + elemH)) <= yThreshold;
+		
+		elem.data('scroll-bounds-callback', function () {
+			App.modules.notify('changeState.update', {
+				item: ctn,
+				state: 'x-start',
+				action: xStartVisible ? 'on' : 'off'
+			});
+			App.modules.notify('changeState.update', {
+				item: ctn,
+				state: 'x-end',
+				action: xEndVisible ? 'on' : 'off'
+			});
+			App.modules.notify('changeState.update', {
+				item: ctn,
+				state: 'y-start',
+				action: yStartVisible ? 'on' : 'off'
+			});
+			App.modules.notify('changeState.update', {
+				item: ctn,
+				state: 'y-end',
+				action: yEndVisible ? 'on' : 'off'
+			});
+		});
+	};
+	
+	var updateData = function (elem) {
+		App.callback(elem.data('scroll-bounds-callback'));
+	};
+	
+	// SCROLL
+	var onElementScroll = function () {
+		var t = $(this);
+		
+		computeStates(t);
```

Comment:
Should we compute the state only on the next frame instead of each events?

Reply:
`computeStates()` only reads, apart the .data call, which should not trash the layout (?does it?) As long as it only reads, it's ok.

Reply:
So nothing changes here